pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py

```diff
@@ -244,7 +252,19 @@ def _blocking_consume(self):
                 break
             except Exception as exc:
                 recover = self._policy.on_exception(exc)
-                if not recover:
+                if recover:
+                    with threading.Lock():
```

Comment:
I am fairly certain that `with threading.Lock():` used in this way does not work, because it is always using a different lock. You need to save a single lock object to the consumer instance so you are calling the `__enter__` method of the same lock.

Basically, `self._recovery_lock = threading.Lock()` in the `__init__.py` file, and then `with self._recovery_lock:` here.

Reply:
What is the issue with always using a different lock?

Reply:
It will never actually lock anything.

Reply:
BLERG

Reply:
Longer explanation: `threading.Lock()` creates a single, discrete lock. To illustrate, imagine a world where we have two separate locks that lock two separate things:

To illustrate, consider a kitchen with two cooks. Both cooks want to make scrambled eggs on the stove top, and they want to put bread in the toaster. Since there is only one stove top and only one toaster, the cooks can not use them at the same time, but it would be fine for one cook to use the stove top while the other one uses the toaster.

These are different lock instances:

```python
stove_lock = threading.Lock()
toaster_lock = threading.Lock()

with stove_lock:
    scramble_eggs()

with toaster_lock:
    toast_bread()
```

When you call `with toaster_lock:`, if another thread is currently executing code within a separate `with toaster_lock:`, then your thread will block until the other thread calls `toaster_lock.__exit__`. However, it does not matter if another thread is currently using `stove_lock`.

But when you just do `with threading.Lock():` then you are creating a _different_ lock instance every time, and it is not actually possible for anything else to be checking against the same lock. The other threads are just blindly using their own, distinct locks and not checking in with yours.